### ISSUE TITLE
Provide GroupAdjacent overload that allows selecting the return enumerable type

### ISSUE THREAD

**Issue by fsateler** [morelinq/MoreLINQ#127]
All `GroupAdjacent` overloads return `IEnumerable<IGrouping<TKey,TResult>>` It would be great to also have one that can return `IEnumerable<TResult>` (ie, pass a `Func<TKey,IEnumerable<TElement>,TResult>` as second parameter).

The current resultSelector overload does not have a matching `GroupBy` overload.

**Comment by atifaziz** [comment]
You can achieve the same result by projecting the groups returned by `GroupAdjacent` via a follow-on `Select`.

**Comment by fsateler** [comment]
First, I take back the part about not having a matching `GroupBy` overload, as [there is one](https://msdn.microsoft.com/en-us/library/bb534304(v=vs.100).aspx), and I was just blind enough to not see it.

Secondly, yes, the same result can be achieved using a follow-on select. However having the overload missing has the following drawbacks:

1. Missing consistency. The more consistency between the MoreLINQ API and the original LINQ api, the better.
2. Performance. Right now, the `elementSelector` is invoked for every element in the source sequence, which yields a new object (with a corresponding memory allocation) and it may not be needed (eg, if the result will be a sum of each group).

Would you have any objections to adding the overload if I wrote a PR?

**Comment by atifaziz** [comment]
Not sure I see how you think invocation of `elementSelector` can avoided, even in the case of summing. Even `Func<TKey, IEnumerable<TElement>, TResult>` would require collecting elements to produce the result.

The only thing I can see that can be saved is allocating a group object on the heap if your `TResult` is a value type.

**Comment by fsateler** [comment]
Collecting elements is still required, but returning intermediate objects for each item in the sequence is not. For example:

    list.GroupAdjacent(i => i.Key, (key, group) => {
        int sum = 0;
        foreach(var item in group) {
               sum += item.Value;
        }
        return new { key, sum };
    });

Note how I saved the indirection via a `Func`, and instead fetch the value directly.

**Comment by atifaziz** [comment]
Sorry if I'm being a little thick lately, but is the step you're looking to save the last `Select` projection that does the sum?

```c#
var ns = Enumerable.Range(1, 10);
var q =
    ns.Interleave(ns, ns, ns)
      .Index()
      .Select(e => new { Key = e.Value, Value = e.Value + e.Key })
      .GroupAdjacent(e => e.Key, e => e.Value)
      .Select(g => new { g.Key, Sum = g.Sum() })
    // could also do...
    //.GroupAdjacent(e => e.Key)
    //.Select(g => new { g.Key, Sum = g.Sum(e => e.Value) });
      ;
```

May be it's just the wording, but not sure what you mean by *fetching the value directly* or the intermediate objects you wish save (the only intermediate I see saving are the grouping objects).

And why wouldn't you just do the following in your example?

```c#
list.GroupAdjacent(i => i.Key, (key, group) => new { key, sum = group.Sum(e => e.Value) });
```

**Comment by fsateler** [comment]
@atifaziz yes, sorry I mixed multiple things in one example. 

> is the step you're looking to save the last Select projection that does the sum?

Yes. Or, more precisely, I'd like to avoid the projection currently done internally by `GroupAdjacent`.

Note that your commented alternative still does the projection. It just happens to be an identity projection.

> May be it's just the wording, but not sure what you mean by fetching the value directly or the intermediate objects you wish save (the only intermediate I see saving are the grouping objects).

Internally, `GroupAdjacent` creates a `List<TElement>`. So that means several allocations for each group. 
If instead I process the source elements, then I would only get the `List<TSource>` bookkeeping overhead (this is what I mean by intermediate objects). This can already be now by passing the identity projection.
Once I have a list of source objects to process, I can calculate my final result without an intermediate `Func<TSource, TElement>` indirection (which probably means the JIT can better optimize). This is what I mean by fetching the value directly.


> And why wouldn't you just do the following in your example?

In the simple `Sum` case it is no problem (except for https://github.com/dotnet/corefx/issues/5549, but that is a separate thing). But for calculating other types of aggregates it may be necessary to implement manually (eg, standard deviation).
But it may be necessary to implement manually if one wants to calculate other types of aggregates (eg, standard deviation), or for returning multiple aggregates (and thus avoid multiple iterations of the source sequence).

**Comment by atifaziz** [comment]
> I'd like to avoid the projection currently done internally by GroupAdjacent.

@fsateler Do you mean the one on [line 91 of GroupAdjacent.cs](https://github.com/morelinq/MoreLINQ/blob/7d1b7fa506339facf61bc6c233b8828af286086b/MoreLinq/GroupAdjacent.cs#L91)?

**Comment by fsateler** [comment]
@atifaziz I'd like to avoid the one [on line 192](https://github.com/morelinq/MoreLINQ/blob/7d1b7fa506339facf61bc6c233b8828af286086b/MoreLinq/GroupAdjacent.cs#L192) if possible.

But, I'd like to pause the performance discussion. I will benchmark the current code and the proposed change to see if the change is meaningful in this regard (it may well not be after #129  was merged, possibly as you noted all the overhead was in the extra `MoveNext`). I will post my results.

The part on consistency has gone a bit unnoticed. I'd still like to have consistency with the [GroupBy method overloads](https://msdn.microsoft.com/en-us/library/system.linq.enumerable.groupby(v=vs.100).aspx). Do you think it would make sense to add the overload, even if it doesn't bring any performance benefits?

**Comment by atifaziz** [comment]
I'm happy to merge the PR if you want to add the overload. In fact, I would've probably added them myself but haven't had burning need so far. I think the consistency argument is good too but then you could also add loads of overloads all over that provide the index of an element in any sequence. However, `Index` seems to solve that problem once & generally so I've refrained. I think in the `GroupAdjacent` case, the overload adds very little complexity & a further generalization of what's already there.

**Comment by fsateler** [comment]
@atifaziz I agree that the indexed overloads do not add much value if we have `Index` :+1: . I will post a PR soonish.

**Comment by fsateler** [comment]
Just for the record, I have just lacked time to finalize this, I have currently a local WIP with the overload but I need to update docs. In the end I did not attempt to eliminate the elementSelector invocation, this can be attempted later if I see a significant performance impact there.

**Comment by atifaziz** [comment]
@fsateler In that case, perhaps a good idea to introduce a “postponed” label, tag this and close for now?

**Comment by fsateler** [comment]
I'm fine with adding a postponed tag, but I don't think closing is appropriate: the issue is neither fixed not resolved as notabug. I plan to finish up the overload and submit the PR as soon as I can, that should be in the coming weeks.

I was just updating the issue to let you know I have not forgotten about this.

**Comment by atifaziz** [comment]
@fsateler Oh, that's fine. No need for the tag then. I'll leave it open and let you submit the PR whenever you can.
